src/connector/src/parser/scalar_adapter.rs

```diff
@@ -104,133 +112,196 @@ impl ToSql for ScalarAdapter<'_> {
-        scalar: ScalarRefImpl<'a>,
+    pub(crate) fn from_scalar(
+        scalar: ScalarRefImpl<'_>,
         ty: &Type,
-    ) -> ConnectorResult<ScalarAdapter<'a>> {
+    ) -> ConnectorResult<ScalarAdapter> {
         Ok(match (scalar, ty, ty.kind()) {
             (ScalarRefImpl::Utf8(s), &Type::UUID, _) => ScalarAdapter::Uuid(s.parse()?),
             (ScalarRefImpl::Utf8(s), &Type::NUMERIC, _) => {
-                ScalarAdapter::Numeric(string_to_pg_numeric(s)?)
+                ScalarAdapter::Numeric(string_to_pg_numeric(s))
             }
             (ScalarRefImpl::Int256(s), &Type::NUMERIC, _) => {
-                ScalarAdapter::Numeric(string_to_pg_numeric(&s.to_string())?)
+                ScalarAdapter::Numeric(string_to_pg_numeric(&s.to_string()))
             }
             (ScalarRefImpl::Utf8(s), _, Kind::Enum(_)) => {
                 ScalarAdapter::Enum(EnumString(s.to_owned()))
             }
-            _ => ScalarAdapter::Builtin(scalar),
+            (ScalarRefImpl::List(list), &Type::NUMERIC_ARRAY, _) => {
+                let mut vec = vec![];
+                for datum in list.iter() {
+                    vec.push(match datum {
+                        Some(ScalarRefImpl::Int256(s)) => Some(string_to_pg_numeric(&s.to_string())),
+                        Some(ScalarRefImpl::Decimal(s)) => Some(rw_numeric_to_pg_numeric(s)),
+                        Some(ScalarRefImpl::Utf8(s)) => Some(string_to_pg_numeric(s)),
+                        None => None,
+                        _ => {
+                            unreachable!("Currently, only rw-numeric[], rw_int256[] and varchar[] are supported to convert to pg-numeric[]");
+                        }
+                    })
+                }
+                ScalarAdapter::NumericList(vec)
+            }
+            (ScalarRefImpl::List(list), _, Kind::Array(inner_type)) => match inner_type.kind() {
+                Kind::Enum(_) => {
+                    let mut vec = vec![];
+                    for datum in list.iter() {
+                        vec.push(match datum {
+                            Some(ScalarRefImpl::Utf8(s)) => Some(EnumString(s.to_owned())),
+                            _ => unreachable!(
+                                "Only non-null varchar[] is supported to convert to enum[]"
+                            ),
+                        })
+                    }
+                    ScalarAdapter::EnumList(vec)
+                }
+                _ => {
+                    let mut vec = vec![];
+                    for datum in list.iter() {
+                        vec.push(
+                            datum
+                                .map(|s| ScalarAdapter::from_scalar(s, inner_type))
+                                .transpose()?,
+                        );
+                    }
+                    ScalarAdapter::List(vec)
+                }
+            },
+            _ => ScalarAdapter::Builtin(scalar.into_scalar_impl()),
```

Comment:
`.into_scalar_impl()` contains a `clone` and is the original reason `ScalarAdapter` contains `ScalarRefImpl`. But given it is only used as primary key during backfill at this time, I guess cloning is okay here.